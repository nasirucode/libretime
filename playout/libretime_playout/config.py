from pathlib import Path
from typing import List, Literal

from libretime_shared.config import (
    BaseConfig,
    GeneralConfig,
    RabbitMQConfig,
    StreamConfig,
)
from pydantic import BaseModel

CACHE_DIR = Path.cwd() / "scheduler"
RECORD_DIR = Path.cwd() / "recorder"

PUSH_INTERVAL = 2
POLL_INTERVAL = 400


class PlayoutConfig(BaseModel):
    liquidsoap_host: str = "localhost"
    liquidsoap_port: int = 1234

    record_file_format: Literal["mp3", "ogg"] = "ogg"  # record_file_type
    record_bitrate: int = 256
    record_samplerate: int = 44100
    record_channels: int = 2
    record_sample_size: int = 16


class LiquidsoapConfig(BaseModel):
    server_listen_address: str = "127.0.0.1"
    server_listen_port: int = 1234

    harbor_listen_address: List[str] = ["0.0.0.0"]


class Config(BaseConfig):
    general: GeneralConfig
    rabbitmq: RabbitMQConfig = RabbitMQConfig()
    playout: PlayoutConfig = PlayoutConfig()
    liquidsoap: LiquidsoapConfig = LiquidsoapConfig()
    stream: StreamConfig = StreamConfig()
