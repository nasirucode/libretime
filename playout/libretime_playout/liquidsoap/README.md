# Liquidsoap

## Inputs pipeline

[![](https://mermaid.ink/svg/pako:eNqdVD1vgzAQ_SvWTY2U_AGGTFkyVAxE6mAyOHAUVLCRfRaqkvz3mgKJoVQJ8fR8H--efWefIVEpQgCfWtQ5O-xiydw6lUJ-8Uhk2MFjZzZJjqktkUc96O0NngxpFBX_GFDvKWRtKcpVw_ctYsZB3_UuCtm7KgedyxMQ6qHOm2kKSnJGig2m1WxgqG8KvJyb7WFSqPeDYi__rn0Jgepge8Q_XK1xNTos22y2U97xrc9HzAqaC_VETrr2VPRIxSVRkrQqzeXO8o8Uj-hBHe_qJsOzJHFBg56n9Vo5Gd4XOV4bo99iypLb8oPq0RHWUKF2zyd17_jcMsdAOVYYQ-BgipmwJcUQy6sLFZZU9C0TCEhbXIOtU0G4K4T7ASoIMlEavP4AY-l5kA)](https://mermaid.live/edit#pako:eNqdVD1vgzAQ_SvWTY2U_AGGTFkyVAxE6mAyOHAUVLCRfRaqkvz3mgKJoVQJ8fR8H--efWefIVEpQgCfWtQ5O-xiydw6lUJ-8Uhk2MFjZzZJjqktkUc96O0NngxpFBX_GFDvKWRtKcpVw_ctYsZB3_UuCtm7KgedyxMQ6qHOm2kKSnJGig2m1WxgqG8KvJyb7WFSqPeDYi__rn0Jgepge8Q_XK1xNTos22y2U97xrc9HzAqaC_VETrr2VPRIxSVRkrQqzeXO8o8Uj-hBHe_qJsOzJHFBg56n9Vo5Gd4XOV4bo99iypLb8oPq0RHWUKF2zyd17_jcMsdAOVYYQ-BgipmwJcUQy6sLFZZU9C0TCEhbXIOtU0G4K4T7ASoIMlEavP4AY-l5kA)
