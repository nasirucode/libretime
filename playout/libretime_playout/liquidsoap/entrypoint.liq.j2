# THIS FILE IS AUTO GENERATED. PLEASE DO NOT EDIT!
###########################################################
# The ignore() lines are to squash unused variable warnings

# Inputs
master_live_stream_mp = "{{ config.stream.inputs.main.mount }}"
master_live_stream_port = {{ config.stream.inputs.main.port }}
dj_live_stream_mp = "{{ config.stream.inputs.show.mount }}"
dj_live_stream_port = {{ config.stream.inputs.show.port }}

{% for output in config.stream.outputs.merged -%}
# Output s{{ loop.index }}
s{{ loop.index }}_enable = {{ output.enabled | lower }}
s{{ loop.index }}_output = "{{ output.kind }}"
s{{ loop.index }}_host = "{{ output.host }}"
s{{ loop.index }}_port = {{ output.port }}
s{{ loop.index }}_mount = "{{ output.mount }}"
s{{ loop.index }}_user = "{{ output.source_user }}"
s{{ loop.index }}_pass = "{{ output.source_password }}"

s{{ loop.index }}_channels = "{{ output.audio.channels.value }}"
s{{ loop.index }}_type = "{{ output.audio.format.value }}"
s{{ loop.index }}_bitrate = {{ output.audio.bitrate }}

s{{ loop.index }}_name = "{{ output.name or '' }}"
s{{ loop.index }}_description = "{{ output.description or '' }}"
s{{ loop.index }}_genre = "{{ output.genre or '' }}"
s{{ loop.index }}_url = "{{ output.website or '' }}"

{% endfor -%}
icecast_vorbis_metadata = {{ icecast_vorbis_metadata | lower }}

# System output
output_sound_device = {{ config.stream.outputs.system[0].enabled | lower }}
output_sound_device_type = "{{ config.stream.outputs.system[0].kind.value }}"

# Settings
auth_path = "{{ paths.auth_filepath }}"

{% if paths.log_filepath is defined -%}
set("log.file.path", "{{ paths.log_filepath }}")
{%- else -%}
set("log.file", false)
{%- endif %}

set("server.telnet", true)
set("server.telnet.bind_addr", "{{ config.liquidsoap.server_listen_address }}")
set("server.telnet.port", {{ config.liquidsoap.server_listen_port }})

set("harbor.bind_addrs", ["{{ config.liquidsoap.harbor_listen_address | join('", "') }}"])

station_name = ref "{{ info.station_name }}"

off_air_meta = ref "{{ preferences.message_offline }}"
stream_metadata_type = ref {{ preferences.message_format.value }}
default_dj_fade = ref {{ preferences.input_fade_transition }}

%include "{{ paths.lib_filepath }}"
