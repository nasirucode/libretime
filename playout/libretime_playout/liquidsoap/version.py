import re
from subprocess import run
from typing import Tuple

LIQUIDSOAP_VERSION_RE = re.compile(r"(?:Liquidsoap )?(\d+).(\d+).(\d+)")
LIQUIDSOAP_MIN_VERSION = (1, 4, 0)


def parse_liquidsoap_version(version: str) -> Tuple[int, int, int]:
    match = LIQUIDSOAP_VERSION_RE.search(version)

    if match is None:
        return (0, 0, 0)
    return (int(match.group(1)), int(match.group(2)), int(match.group(3)))


def get_liquidsoap_version() -> Tuple[int, int, int]:
    cmd = run(
        ("liquidsoap", "--check", "print(liquidsoap.version) shutdown()"),
        check=True,
        capture_output=True,
        text=True,
    )

    return parse_liquidsoap_version(cmd.stdout)
