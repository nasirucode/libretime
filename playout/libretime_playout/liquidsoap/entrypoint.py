from pathlib import Path
from typing import Optional, Tuple

from jinja2 import Template
from libretime_shared.config import AudioFormat, IcecastOutput, SystemOutput

from ..config import Config
from .models import Info, StreamPreferences

here = Path(__file__).parent

entrypoint_template_path = here / "entrypoint.liq.j2"
entrypoint_template = Template(
    entrypoint_template_path.read_text(encoding="utf-8"),
    keep_trailing_newline=True,
)

# Liquidsoap has 4 hardcoded output stream set of variables, so we need to
# fill the missing stream outputs with placeholders so Liquidsoap does
# not fail with missing variables in the entrypoint.
_icecast_placeholder = IcecastOutput(
    enabled=False,
    mount="",
    source_password="",
    audio=dict(format="ogg", bitrate=256),
)

_system_placeholder = SystemOutput()


def generate_entrypoint(
    entrypoint_filepath: Path,
    log_filepath: Optional[Path],
    config: Config,
    preferences: StreamPreferences,
    info: Info,
    version: Tuple[int, int, int],
):
    paths = {}
    paths["auth_filepath"] = here / "liquidsoap_auth.py"
    paths["lib_filepath"] = here / f"{version[0]}.{version[1]}/ls_script.liq"

    if log_filepath is not None:
        paths["log_filepath"] = log_filepath.resolve()

    config = config.copy()
    missing_outputs = [_icecast_placeholder] * (4 - len(config.stream.outputs.merged))
    config.stream.outputs.icecast.extend(missing_outputs)

    if not config.stream.outputs.system:
        config.stream.outputs.system.append(_system_placeholder)

    # Global icecast_vorbis_metadata until it is
    # handled per output
    icecast_vorbis_metadata = any(
        o.enabled and o.audio.format == AudioFormat.OGG and o.audio.enable_metadata  # type: ignore
        for o in config.stream.outputs.icecast
    )

    entrypoint_filepath.write_text(
        entrypoint_template.render(
            config=config,
            preferences=preferences,
            info=info,
            paths=paths,
            version=version,
            icecast_vorbis_metadata=icecast_vorbis_metadata,
        ),
        encoding="utf-8",
    )
