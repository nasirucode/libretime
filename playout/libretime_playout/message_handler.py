import json
from queue import Queue as ThreadQueue
from signal import SIGTERM, signal
from time import sleep
from typing import Any, Dict

# For RabbitMQ
from kombu.connection import Connection
from kombu.messaging import Exchange, Queue
from kombu.mixins import ConsumerMixin
from loguru import logger

from .config import Config


class MessageHandler(ConsumerMixin):
    def __init__(
        self,
        connection: Connection,
        fetch_queue: ThreadQueue,
        recorder_queue: ThreadQueue,
    ):
        self.connection = connection

        self.fetch_queue = fetch_queue
        self.recorder_queue = recorder_queue

    def get_consumers(self, Consumer, channel):
        exchange = Exchange("airtime-pypo", "direct", durable=True, auto_delete=True)
        queues = [Queue("pypo-fetch", exchange=exchange, key="foo")]

        return [
            Consumer(queues, callbacks=[self.on_message], accept=["text/plain"]),
        ]

    def on_message(self, body, message):
        logger.debug(f"received message: {body}")
        try:
            try:
                body = body.decode()
            except (UnicodeDecodeError, AttributeError):
                pass

            payload = json.loads(body)
            command = payload["event_type"]
            logger.info(f"handling command: {command}")

            if command in (
                "update_schedule",
                "reset_liquidsoap_bootstrap",
                "update_stream_format",
                "update_message_offline",
                "update_station_name",
                "switch_source",
                "update_transition_fade",
                "disconnect_source",
            ):
                self.fetch_queue.put(message.payload)

            elif command in (
                "update_recorder_schedule",
                "cancel_recording",
            ):
                self.recorder_queue.put(message.payload)

            else:
                logger.warning(f"invalid command: {command}")

        except Exception as exception:
            logger.exception(exception)

        message.ack()


class MessageListener:
    def __init__(
        self,
        config: Config,
        fetch_queue: ThreadQueue[Dict[str, Any]],
        recorder_queue: ThreadQueue[Dict[str, Any]],
    ) -> None:
        self.config = config

        self.fetch_queue = fetch_queue
        self.recorder_queue = recorder_queue

    def run_forever(self):
        while True:
            with Connection(
                self.config.rabbitmq.url,
                heartbeat=5,
                transport_options={"client_properties": {"connection_name": "playout"}},
            ) as connection:
                handler = MessageHandler(
                    connection=connection,
                    fetch_queue=self.fetch_queue,
                    recorder_queue=self.recorder_queue,
                )

                def shutdown(_signum, _frame):
                    raise SystemExit()

                signal(SIGTERM, shutdown)

                logger.info("starting message handler")
                handler.run()

            sleep(5)
