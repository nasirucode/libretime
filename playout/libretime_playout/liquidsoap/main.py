import os
from pathlib import Path
from typing import Optional

import click
from libretime_api_client.v2 import ApiClient
from libretime_shared.cli import cli_config_options, cli_logging_options
from libretime_shared.config import DEFAULT_ENV_PREFIX
from libretime_shared.logging import level_from_name, setup_logger
from loguru import logger

from ..config import Config
from .entrypoint import generate_entrypoint
from .models import Info, StreamPreferences
from .version import get_liquidsoap_version

here = Path(__file__).parent


@click.command(context_settings={"auto_envvar_prefix": DEFAULT_ENV_PREFIX})
@cli_logging_options()
@cli_config_options()
def cli(log_level: str, log_filepath: Optional[Path], config_filepath: Optional[Path]):
    """
    Run liquidsoap.
    """
    logger_level, _ = setup_logger(level_from_name(log_level), log_filepath)
    config = Config(config_filepath)

    api_client = ApiClient(
        base_url=config.general.public_url,
        api_key=config.general.api_key,
    )

    version = get_liquidsoap_version()

    info = Info(**api_client.get_info().json())
    preferences = StreamPreferences(**api_client.get_stream_preferences().json())

    entrypoint_filepath = Path.cwd() / "radio.liq"
    generate_entrypoint(
        entrypoint_filepath,
        log_filepath,
        config,
        preferences,
        info,
        version,
    )

    exec_args = [
        "/usr/bin/liquidsoap",
        "libretime-liquidsoap",
        "--verbose",
        str(entrypoint_filepath),
    ]
    if logger_level.is_debug():
        exec_args.append("--debug")

    logger.debug(f"liquidsoap {version} using script: {entrypoint_filepath}")
    os.execl(*exec_args)
