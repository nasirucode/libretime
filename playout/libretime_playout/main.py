"""
Python part of radio playout (pypo)
"""

import sys
import time
from datetime import datetime
from pathlib import Path
from queue import Queue
from typing import Any, Dict, Optional

import click
from libretime_api_client.v1 import ApiClient as LegacyClient
from libretime_api_client.v2 import ApiClient
from libretime_shared.cli import cli_config_options, cli_logging_options
from libretime_shared.config import DEFAULT_ENV_PREFIX
from libretime_shared.logging import level_from_name, setup_logger
from loguru import logger

from .config import CACHE_DIR, RECORD_DIR, Config
from .history.stats import StatsCollectorThread
from .liquidsoap.client import LiquidsoapClient
from .liquidsoap.version import LIQUIDSOAP_MIN_VERSION
from .message_handler import MessageListener
from .player.fetch import PypoFetch
from .player.file import PypoFile
from .player.liquidsoap import PypoLiquidsoap
from .player.push import PypoPush
from .recorder import Recorder


@click.command(context_settings={"auto_envvar_prefix": DEFAULT_ENV_PREFIX})
@cli_logging_options()
@cli_config_options()
def cli(log_level: str, log_filepath: Optional[Path], config_filepath: Optional[Path]):
    """
    Run playout.
    """
    setup_logger(level_from_name(log_level), log_filepath)
    config = Config(config_filepath)

    try:
        for dir_path in [CACHE_DIR, RECORD_DIR]:
            dir_path.mkdir(exist_ok=True)
    except OSError as exception:
        logger.error(exception)
        sys.exit(1)

    logger.info("###########################################")
    logger.info("#             *** pypo  ***               #")
    logger.info("#   Liquidsoap Scheduled Playout System   #")
    logger.info("###########################################")

    # Although all of our calculations are in UTC, it is useful to know what timezone
    # the local machine is, so that we have a reference for what time the actual
    # log entries were made
    logger.info("Timezone: %s" % str(time.tzname))
    logger.info("UTC time: %s" % str(datetime.utcnow()))

    legacy_client = LegacyClient()
    api_client = ApiClient(
        base_url=config.general.public_url,
        api_key=config.general.api_key,
    )

    while not legacy_client.is_server_compatible():
        time.sleep(5)

    success = False
    while not success:
        try:
            legacy_client.register_component("pypo")
            success = True
        except Exception as exception:
            logger.exception(exception)
            time.sleep(10)

    liq_client = LiquidsoapClient(
        host=config.playout.liquidsoap_host,
        port=config.playout.liquidsoap_port,
    )

    logger.debug("Checking if Liquidsoap is running")
    liq_version = liq_client.wait_for_version()
    if not LIQUIDSOAP_MIN_VERSION <= liq_version:
        raise Exception(f"Invalid liquidsoap version {liq_version}")

    fetch_queue: Queue[Dict[str, Any]] = Queue()
    recorder_queue: Queue[Dict[str, Any]] = Queue()
    push_queue: Queue[Dict[str, Any]] = Queue()
    # This queue is shared between pypo-fetch and pypo-file, where pypo-file
    # is the consumer. Pypo-fetch will send every schedule it gets to pypo-file
    # and pypo will parse this schedule to determine which file has the highest
    # priority, and retrieve it.
    file_queue: Queue[Dict[str, Any]] = Queue()

    pypo_liquidsoap = PypoLiquidsoap(liq_client)

    file_thread = PypoFile(file_queue, api_client)
    file_thread.start()

    fetch_thread = PypoFetch(
        fetch_queue,
        push_queue,
        file_queue,
        liq_client,
        pypo_liquidsoap,
        config,
        api_client,
        legacy_client,
    )
    fetch_thread.start()

    push_thread = PypoPush(push_queue, pypo_liquidsoap, config)
    push_thread.start()

    recorder_thread = Recorder(recorder_queue, config, legacy_client)
    recorder_thread.start()

    stats_collector_thread = StatsCollectorThread(config, legacy_client)
    stats_collector_thread.start()

    message_listener = MessageListener(config, fetch_queue, recorder_queue)
    message_listener.run_forever()
