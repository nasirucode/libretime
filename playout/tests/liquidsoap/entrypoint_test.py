from pathlib import Path
from unittest import mock

import pytest

from libretime_playout.config import Config
from libretime_playout.liquidsoap.entrypoint import generate_entrypoint
from libretime_playout.liquidsoap.models import Info, StreamPreferences


@pytest.mark.parametrize(
    "version",
    [pytest.param((1, 4, 4), id="1.4")],
)
def test_generate_entrypoint(tmp_path: Path, config: Config, version, snapshot):
    entrypoint_filepath = tmp_path / "radio.liq"

    with mock.patch(
        "libretime_playout.liquidsoap.entrypoint.here",
        Path("/fake"),
    ):
        generate_entrypoint(
            entrypoint_filepath,
            log_filepath=Path("/var/log/radio.log"),
            config=config,
            preferences=StreamPreferences(
                input_fade_transition=0.0,
                message_format=0,
                message_offline="LibreTime - offline",
            ),
            info=Info(
                station_name="LibreTime",
            ),
            version=version,
        )

    found = entrypoint_filepath.read_text(encoding="utf-8")
    assert found == snapshot
