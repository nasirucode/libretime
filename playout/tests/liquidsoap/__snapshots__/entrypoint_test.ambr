# name: test_generate_entrypoint[1.4]
  '''
  # THIS FILE IS AUTO GENERATED. PLEASE DO NOT EDIT!
  ###########################################################
  # The ignore() lines are to squash unused variable warnings
  
  # Inputs
  master_live_stream_mp = "main"
  master_live_stream_port = 8001
  dj_live_stream_mp = "show"
  dj_live_stream_port = 8002
  
  # Output s1
  s1_enable = true
  s1_output = "icecast"
  s1_host = "localhost"
  s1_port = 8000
  s1_mount = "main"
  s1_user = "source"
  s1_pass = "hackme"
  
  s1_channels = "stereo"
  s1_type = "ogg"
  s1_bitrate = 256
  
  s1_name = "LibreTime!"
  s1_description = "LibreTime Radio! Stream #1"
  s1_genre = "various"
  s1_url = "https://libretime.org"
  
  # Output s2
  s2_enable = true
  s2_output = "icecast"
  s2_host = "localhost"
  s2_port = 8000
  s2_mount = "second"
  s2_user = "source"
  s2_pass = "hackme"
  
  s2_channels = "stereo"
  s2_type = "mp3"
  s2_bitrate = 256
  
  s2_name = ""
  s2_description = ""
  s2_genre = ""
  s2_url = ""
  
  # Output s3
  s3_enable = false
  s3_output = "icecast"
  s3_host = "localhost"
  s3_port = 8000
  s3_mount = ""
  s3_user = "source"
  s3_pass = ""
  
  s3_channels = "stereo"
  s3_type = "ogg"
  s3_bitrate = 256
  
  s3_name = ""
  s3_description = ""
  s3_genre = ""
  s3_url = ""
  
  # Output s4
  s4_enable = false
  s4_output = "icecast"
  s4_host = "localhost"
  s4_port = 8000
  s4_mount = ""
  s4_user = "source"
  s4_pass = ""
  
  s4_channels = "stereo"
  s4_type = "ogg"
  s4_bitrate = 256
  
  s4_name = ""
  s4_description = ""
  s4_genre = ""
  s4_url = ""
  
  icecast_vorbis_metadata = false
  
  # System output
  output_sound_device = false
  output_sound_device_type = "alsa"
  
  # Settings
  auth_path = "/fake/liquidsoap_auth.py"
  
  set("log.file.path", "/var/log/radio.log")
  
  set("server.telnet", true)
  set("server.telnet.bind_addr", "127.0.0.1")
  set("server.telnet.port", 1234)
  
  set("harbor.bind_addrs", ["0.0.0.0"])
  
  station_name = ref "LibreTime"
  
  off_air_meta = ref "LibreTime - offline"
  stream_metadata_type = ref 0
  default_dj_fade = ref 0.0
  
  %include "/fake/1.4/ls_script.liq"
  
  '''
# ---
